# STAT 207 Group Lab Assignment 10 - [10 total points]

## Multiple Linear Regression: Test Data, Interactions, & Assumptions

<hr>

## <u>Lab Grading</u>:

Should we grade your submission?  If not, write the netID of the submission to be graded.  (Note: We will only grade one assignment per group, and we'll pick the first one that says we should grade that submission.  We will assign the same grade to all team members.)

*For example*, you might respond: **grade this submission** or **my submission is under netID jdeeke**

mdm16


If you said **my submission is under netID** above, we will not read any more of your lab submission.

If you said **grade this submission** above, who worked with you on this submission?  Write both their **names** and **netIDs**.  Also, discuss and record who from your group can drive stick shift (manual).  If so, who taught you?  If not, would you want to learn how to drive stick?

<hr>

## <u>Purpose</u>:
You should work in groups of 2-3 on this report (not working in groups without permission will result in a point deduction). The purpose of this group lab assignment is to incorporate an interaction term into a multiple linear regression model and to assess the model fit.  

<hr>

## <u>Assignment Instructions</u>:

### Group Roles

Suggested and specified roles are provided below: 

#### Groups of 2

* **Driver**: This student will type the report.  While typing the report, you may be the one who is selecting the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  Specific ways to help may include: outlining the general steps needed to solve a question (providing the overview), locating examples within the course notes, and reviewing each line of code as it is typed.

#### Groups of 3

* **Driver**: This student will type the report.  They may also be the one to select the functions to apply to the data.
* **Navigator**: This student will guide the process of answering the question.  They may select the general approach to answering the question and/or a few steps to be completed along the way. 
* **Communicator**: This student will review the report (as it is typed) to ensure that it is clear and concise.  This student may also locate relevant examples within the course notes that may help complete the assignment.

<hr>

### Imports

In [1]:
#Run this
import pandas as pd                    # imports pandas and calls the imported version 'pd'
import matplotlib.pyplot as plt        # imports the package and calls it 'plt'
import seaborn as sns                  # imports the seaborn package with the imported name 'sns'
sns.set()  
from sklearn.linear_model import LinearRegression
import statsmodels.formula.api as smf

## Case Study: Car Sales Data

Recall that we worked with car sales data in Lab 9.  We will continue working with this same data for this lab.  The car sales dataset contains the following variables for used cars listed for sale from cardekho.com:

- **name**: the type of car for sale
- **year**: the year in which the car was bought
- **selling_price**: the listed selling price of the car (in rupees; currently 1 rupee = 0.012 USD)
- **km_driven**: the total mileage of the car, in kilometers
- **fuel**: the type of fuel needed for the car
- **seller_type**: reports if the car is sold by an individual or a dealer
- **transmission**: gear transmission of the car (Automatic/Manual)
- **owner**: number of previous owners of the car
- **mileage_kmpl**: the fuel efficiency mileage, in kilometers per Liter
- **engine_CC**: the engine capacity of the car, in cubic centimeters
- **max_power_bhp**: the maximum power of the engine, in brake horse power
- **seats**: the number of seats in the car.

Use the car sales dataset in this folder (car_sales.csv) to perform the following analyses.

The code cell below will read in the data for you.  Be sure to run the cell. 

In [2]:
df = pd.read_csv('car_sales.csv', na_values = [""])
df = df.dropna()
df.head()

,name,year,selling_price,km_driven,fuel,seller_type,transmission,owner,mileage_kmpl,engine_CC,max_power_bhp,seats
0,Maruti Swift Dzire VDI,2014,450000,145500,Diesel,Individual,Manual,First Owner,23.40,1248.0,74.00,5.0
1,Skoda Rapid 1.5 TDI Ambition,2014,370000,120000,Diesel,Individual,Manual,Second Owner,21.14,1498.0,103.52,5.0
2,Honda City 2017-2020 EXi,2006,158000,140000,Petrol,Individual,Manual,Third Owner,17.70,1497.0,78.00,5.0
3,Hyundai i20 Sportz Diesel,2010,225000,127000,Diesel,Individual,Manual,First Owner,23.00,1396.0,90.00,5.0
4,Maruti Swift VXI BSIII,2007,130000,120000,Petrol,Individual,Manual,First Owner,16.10,1298.0,88.20,5.0


### 1. [2 point] Fitting a Linear Model

We'll aim to predict the selling price of a car.

**a)** Before fitting our model, we'll set aside a portion of the data to serve as a test set.  This way, we can help to ensure that our model is not overfit to our training data, capturing too much noise that is specific to our data.

To start, set aside 20% of the data (using a random_state of 358).

In [3]:
test_df = df.sample(frac=0.2, random_state=358)
train_df = df.drop(test_df.index)


**b)**  For this model, select three quantitative variables and one categorical variable to serve as your four predictor variables.  

Then, fit a linear model to predict the selling price of a car from your selected quantitative and categorical variables, along with an interaction between one of your quantitative variables and your categorical variable.  Be sure to fit the model with the appropriate data set.

In [4]:
car_model = smf.ols(
    'selling_price ~ year + km_driven + max_power_bhp * transmission',
    data=train_df
).fit()

car_model.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:          selling_price   R-squared:                       0.774
Model:                            OLS   Adj. R-squared:                  0.774
Method:                 Least Squares   F-statistic:                     4338.
Date:                Fri, 03 Apr 2026   Prob (F-statistic):               0.00
Time:                        11:27:07   Log-Likelihood:                -90351.
No. Observations:                6325   AIC:                         1.807e+05
Df Residuals:                    6319   BIC:                         1.808e+05
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
========================================================================================================
                                           coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------------------
Intercept                            -1.074e+08   2.89e+06    -37.211      0.000   -1.13e+08   -1.02e+08
transmission[T.Manual]                1.722e+06   4.38e+04     39.331      0.000    1.64e+06    1.81e+06
year                                   5.25e+04   1430.766     36.693      0.000    4.97e+04    5.53e+04
km_driven                               -0.4248      0.092     -4.596      0.000      -0.606      -0.244
max_power_bhp                         2.451e+04    258.249     94.925      0.000     2.4e+04     2.5e+04
max_power_bhp:transmission[T.Manual] -1.858e+04    334.797    -55.482      0.000   -1.92e+04   -1.79e+04
==============================================================================
Omnibus:                     1886.016   Durbin-Watson:                   1.715
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            96304.782
Skew:                           0.639   Prob(JB):                         0.00
Kurtosis:                      22.073   Cond. No.                     5.42e+07
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 5.42e+07. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### 2. [2.5 points] Interpret Your Model

We will interpret the model from Question 1 below.

**a)** What is the baseline level for the categorical variable?

**b)** Write the fitted model for a specific non-baseline level of your categorical variable.  You may pick the non-baseline level that you would like to explore.

**c)** What is the difference in intercepts between the non-baseline level and the baseline level selected in **part b**?  What is the difference in slopes for your quantitative variable from the interaction term between the non-baseline level and the baseline level selected in **part b**?

### 3. [3.5 points] Evaluate Your Model Fit

For this question, we will explore if our model is appropriate for the data.

**a)** First, generate a fitted vs. residual plot for the training data, which allows us to determine if a required assumption is met for the linear model to be appropriate to model the data.

**b)** Based on this graph, is the model an appropriate way to represent and summarize the relationship between your chosen variables?  Explain.

**c)** Calculate the RMSE on your training data.  Then, calculate the RMSE on your test data.

**d)** Which of the RMSEs would be the best estimate for the typical size of an error for the selling price of a car that was just posted to cardekho.com.  Briefly explain.

### 4. [2 points] Check for Multicollinearity

**a)** Check the model for multicollinearity among the quantitative predictors in your model within your training data.  There are a couple of ways to check the model, and you may choose which to use.

**b)** Based on what you see from part a, determine if or how you might want to change your model.  Briefly explain any changes that you'd like to make to the model.  No need to implement your suggested changes.

### 5. AI Acknowledgement

Our course policy is that you should write all of your own interpretations and other narrative answers (phrases or sentences) yourself without the assistance of AI.  AI should not be used on Questions 2, 3b, 3d, and 4b of this assignment.  

You may use AI to help guide your code, although you should write all of your own code yourself (not copy-paste from another source) and you should cite your use of AI.  I would encourage you to try to generate any necessary code yourself first using course resources and using AI as a debugging tool if/when you reach an error that you can't figure out or to help you perform any coding tasks that are more advanced than we've demonstrated during class (intended only for projects).  

Did you use AI on this assignment?  Did you use other resources outside of our course-provided resources on this assignment?

If you used AI or other resources, answer the following questions to cite your usage.

- Which AI and/or resources did you use?
- What prompts did you ask it?
- How did you integrate the responses into your assignment?  Specifically, which questions or parts are associated with this usage?

Note: answering these three questions are enough for our course but may not be enough for a different course or context.